In this project, we will build a machine-learning system used to detect and predict fraudulent credit card transactions from a dataset consisting of verified and fraudulent credit card transactions. We will split the data and train it with our machine-learning model, in this case, we will be utilizing the Logistic Regression algorithm imported from the scikit-learn library to make our predictions. 

In [2]:
#import libraries
import numpy as np
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import accuracy_score


In [5]:
#load dataset to a dataframe
df = pd.read_csv("creditcard.csv")

In [11]:
#print first 15 rows
df.head()

Time        V1        V2        V3        V4        V5        V6        V7  \
0   0.0 -1.359807 -0.072781  2.536347  1.378155 -0.338321  0.462388  0.239599   
1   0.0  1.191857  0.266151  0.166480  0.448154  0.060018 -0.082361 -0.078803   
2   1.0 -1.358354 -1.340163  1.773209  0.379780 -0.503198  1.800499  0.791461   
3   1.0 -0.966272 -0.185226  1.792993 -0.863291 -0.010309  1.247203  0.237609   
4   2.0 -1.158233  0.877737  1.548718  0.403034 -0.407193  0.095921  0.592941   

         V8        V9  ...       V21       V22       V23       V24       V25  \
0  0.098698  0.363787  ... -0.018307  0.277838 -0.110474  0.066928  0.128539   
1  0.085102 -0.255425  ... -0.225775 -0.638672  0.101288 -0.339846  0.167170   
2  0.247676 -1.514654  ...  0.247998  0.771679  0.909412 -0.689281 -0.327642   
3  0.377436 -1.387024  ... -0.108300  0.005274 -0.190321 -1.175575  0.647376   
4 -0.270533  0.817739  ... -0.009431  0.798278 -0.137458  0.141267 -0.206010   

        V26       V27       V28  Amount  Class  
0 -0.189115  0.133558 -0.021053  149.62      0  
1  0.125895 -0.008983  0.014724    2.69      0  
2 -0.139097 -0.055353 -0.059752  378.66      0  
3 -0.221929  0.062723  0.061458  123.50      0  
4  0.502292  0.219422  0.215153   69.99      0  

[5 rows x 31 columns]

We can see that:
-The amount of each trnsaction is in USD
-Class rerpresents 0 (verified) & 1 (fraud) transaction type
-V1 through V28 are features of each transaction

In [7]:
#get data info
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 284807 entries, 0 to 284806
Data columns (total 31 columns):
 #   Column  Non-Null Count   Dtype  
---  ------  --------------   -----  
 0   Time    284807 non-null  float64
 1   V1      284807 non-null  float64
 2   V2      284807 non-null  float64
 3   V3      284807 non-null  float64
 4   V4      284807 non-null  float64
 5   V5      284807 non-null  float64
 6   V6      284807 non-null  float64
 7   V7      284807 non-null  float64
 8   V8      284807 non-null  float64
 9   V9      284807 non-null  float64
 10  V10     284807 non-null  float64
 11  V11     284807 non-null  float64
 12  V12     284807 non-null  float64
 13  V13     284807 non-null  float64
 14  V14     284807 non-null  float64
 15  V15     284807 non-null  float64
 16  V16     284807 non-null  float64
 17  V17     284807 non-null  float64
 18  V18     284807 non-null  float64
 19  V19     284807 non-null  float64
 20  V20     284807 non-null  float64
 21  V21     28

Here we see there are 284,807 entries in this dataset.

In [8]:
#check # of missing values in columns
df.isnull().sum()

Time      0
V1        0
V2        0
V3        0
V4        0
V5        0
V6        0
V7        0
V8        0
V9        0
V10       0
V11       0
V12       0
V13       0
V14       0
V15       0
V16       0
V17       0
V18       0
V19       0
V20       0
V21       0
V22       0
V23       0
V24       0
V25       0
V26       0
V27       0
V28       0
Amount    0
Class     0
dtype: int64

Results in no values missing. 

In [14]:
#check the number of verified and fraudulent transactions
df["Class"].value_counts()

0    284315
1       492
Name: Class, dtype: int64

The distributin is very unbalanced, with there being 284,315 verified transactions and 492 fraudulent transactions. We want to separate the transactions into two variables, verified and fradulent. 

In [18]:
#separate transaction types
verified = df[df.Class==0]
fraud = df[df.Class==1]

In [19]:
#print values
print(verified.shape)
print(fraud.shape)

(284315, 31)
(492, 31)


In [20]:
#print statistical measures of data for verified transactions
verified.Amount.describe()

count    284315.000000
mean         88.291022
std         250.105092
min           0.000000
25%           5.650000
50%          22.000000
75%          77.050000
max       25691.160000
Name: Amount, dtype: float64

We can see that:
-Mean amount spent is $88
-Standard deviation of data is 250.10592
-Percentages of amount of the purchase 

In [21]:
#print statistical measures of data for fraudulent transactions
fraud.Amount.describe()

count     492.000000
mean      122.211321
std       256.683288
min         0.000000
25%         1.000000
50%         9.250000
75%       105.890000
max      2125.870000
Name: Amount, dtype: float64

-Mean amount spent is $122
-Standard deviation of data is 256.683288
-Percentages of amount of the purchase is much different compared to verified transactions.

In [22]:
#compare values for verified and fraud transactions
df.groupby("Class").mean()

Time        V1        V2        V3        V4        V5  \
Class                                                                   
0      94838.202258  0.008258 -0.006271  0.012171 -0.007860  0.005453   
1      80746.806911 -4.771948  3.623778 -7.033281  4.542029 -3.151225   

             V6        V7        V8        V9  ...       V20       V21  \
Class                                          ...                       
0      0.002419  0.009637 -0.000987  0.004467  ... -0.000644 -0.001235   
1     -1.397737 -5.568731  0.570636 -2.581123  ...  0.372319  0.713588   

            V22       V23       V24       V25       V26       V27       V28  \
Class                                                                         
0     -0.000024  0.000070  0.000182 -0.000072 -0.000089 -0.000295 -0.000131   
1      0.014049 -0.040308 -0.105130  0.041449  0.051648  0.170575  0.075667   

           Amount  
Class              
0       88.291022  
1      122.211321  

[2 rows x 30 columns]

We can see the difference in the means from each transaction type. 

Since we have an imbalanced amount of data, we want to take a random sample of about the same number of verified transactions as there are fraudulent transactions using the .sample function.

In [33]:
#take smaller sample of verified transactions
verifiedSample= verified.sample(n=492)

In [34]:
#concat our sample and fraudulent transactions dataframe
df2 = pd.concat([verifiedSample, fraud], axis=0)

In [35]:
#print first 5 rows
df2.head()

Time        V1        V2        V3        V4        V5        V6  \
3449     2939.0 -0.458755  0.605384  1.176174 -2.150163  0.300584 -0.830793   
89191   62468.0 -0.517493  0.826563  0.811917 -0.102301  1.149607 -0.175053   
70101   53709.0  1.130941  0.274630  0.542070  0.937542 -0.109606 -0.084768   
87986   61931.0  1.254682  0.158501  0.247629  0.718493 -0.610381 -1.134508   
112692  72767.0 -4.578540  1.023057 -2.294984  1.503353 -2.666903 -1.371363   

              V7        V8        V9  ...       V21       V22       V23  \
3449    0.984875 -0.159913  0.584435  ...  0.151025  0.803849 -0.251524   
89191   0.737673 -0.052556 -0.424828  ...  0.118364  0.355665 -0.280789   
70101  -0.026067  0.081163 -0.298001  ... -0.141155 -0.403601  0.107511   
87986  -0.096711 -0.049951  0.442108  ... -0.319031 -1.052264  0.161128   
112692 -1.592653  2.533262 -0.559691  ...  0.290586 -0.189467 -0.537756   

             V24       V25       V26       V27       V28  Amount  Class  
3449    0.224740 -0.063281 -0.839213  0.250582 -0.025604    1.00      0  
89191  -0.741436  0.126849 -0.463532 -0.048377  0.054531    2.00      0  
70101  -0.017835  0.281269 -0.620396  0.039441  0.018328    9.99      0  
87986   0.265814  0.125359  0.103804 -0.040750  0.024964    1.79      0  
112692  0.761514 -1.245135 -0.816001 -1.144781 -0.597719   93.77      0  

[5 rows x 31 columns]

The first 5 rows are all verified transactions as the class is 0.

In [36]:
#p[rint last 5 rows
df2.tail()

Time        V1        V2        V3        V4        V5        V6  \
279863  169142.0 -1.927883  1.125653 -4.518331  1.749293 -1.566487 -2.010494   
280143  169347.0  1.378559  1.289381 -5.004247  1.411850  0.442581 -1.326536   
280149  169351.0 -0.676143  1.126366 -2.213700  0.468308 -1.120541 -0.003346   
281144  169966.0 -3.113832  0.585864 -5.399730  1.817092 -0.840618 -2.943548   
281674  170348.0  1.991976  0.158476 -2.583441  0.408670  1.151147 -0.096695   

              V7        V8        V9  ...       V21       V22       V23  \
279863 -0.882850  0.697211 -2.064945  ...  0.778584 -0.319189  0.639419   
280143 -1.413170  0.248525 -1.127396  ...  0.370612  0.028234 -0.145640   
280149 -2.234739  1.210158 -0.652250  ...  0.751826  0.834108  0.190944   
281144 -2.208002  1.058733 -1.632333  ...  0.583276 -0.269209 -0.456108   
281674  0.223050 -0.068384  0.577829  ... -0.164350 -0.295135 -0.072173   

             V24       V25       V26       V27       V28  Amount  Class  
279863 -0.294885  0.537503  0.788395  0.292680  0.147968  390.00      1  
280143 -0.081049  0.521875  0.739467  0.389152  0.186637    0.76      1  
280149  0.032070 -0.739695  0.471111  0.385107  0.194361   77.89      1  
281144 -0.183659 -0.328168  0.606116  0.884876 -0.253700  245.00      1  
281674 -0.450261  0.313267 -0.289617  0.002988 -0.015309   42.53      1  

[5 rows x 31 columns]

The last 5 rows are all fraudulent transactions as the class is 1.

In [37]:
#check the amount of values in new data frame
df2["Class"].value_counts()

0    492
1    492
Name: Class, dtype: int64

In [38]:
#compare values forverified and fraud transactions
df2.groupby("Class").mean()

Time        V1        V2        V3        V4        V5  \
Class                                                                   
0      94489.540650 -0.027239  0.100850  0.058730 -0.100963 -0.028992   
1      80746.806911 -4.771948  3.623778 -7.033281  4.542029 -3.151225   

             V6        V7        V8        V9  ...       V20       V21  \
Class                                          ...                       
0     -0.031815 -0.030982  0.002947  0.078438  ...  0.050330  0.054228   
1     -1.397737 -5.568731  0.570636 -2.581123  ...  0.372319  0.713588   

            V22       V23      V24       V25       V26       V27       V28  \
Class                                                                        
0     -0.059837 -0.035190  0.02817 -0.019892 -0.011368  0.023745 -0.000385   
1      0.014049 -0.040308 -0.10513  0.041449  0.051648  0.170575  0.075667   

           Amount  
Class              
0       85.145142  
1      122.211321  

[2 rows x 30 columns]

Now that we have good sample data, we can split the data so we can train our machine-learning model.

In [40]:
#X is our target data, we will drop the "class" column
#Y is our features data (V1, V2, V3, etc.)
X = df2.drop(columns="Class", axis=1)
Y = df2["Class"]

In [42]:
print(X)

            Time        V1        V2        V3        V4        V5        V6  \
3449      2939.0 -0.458755  0.605384  1.176174 -2.150163  0.300584 -0.830793   
89191    62468.0 -0.517493  0.826563  0.811917 -0.102301  1.149607 -0.175053   
70101    53709.0  1.130941  0.274630  0.542070  0.937542 -0.109606 -0.084768   
87986    61931.0  1.254682  0.158501  0.247629  0.718493 -0.610381 -1.134508   
112692   72767.0 -4.578540  1.023057 -2.294984  1.503353 -2.666903 -1.371363   
...          ...       ...       ...       ...       ...       ...       ...   
279863  169142.0 -1.927883  1.125653 -4.518331  1.749293 -1.566487 -2.010494   
280143  169347.0  1.378559  1.289381 -5.004247  1.411850  0.442581 -1.326536   
280149  169351.0 -0.676143  1.126366 -2.213700  0.468308 -1.120541 -0.003346   
281144  169966.0 -3.113832  0.585864 -5.399730  1.817092 -0.840618 -2.943548   
281674  170348.0  1.991976  0.158476 -2.583441  0.408670  1.151147 -0.096695   

              V7        V8        V9  .

In [43]:
print(Y)

3449      0
89191     0
70101     0
87986     0
112692    0
         ..
279863    1
280143    1
280149    1
281144    1
281674    1
Name: Class, Length: 984, dtype: int64


Split data into our training data and our testing data with a 80(train), 20(test) ratio using the train_test_split function. 

In [50]:
#assign our X test/train and Y test/train by splitting the data 80:20
X_train, X_test, Y_train, Y_test = train_test_split(X, Y, test_size=0.2)

In [51]:
#print the amount of data for each variable after the split
print(X_train.shape, X_test.shape)
print(Y_train.shape, Y_test.shape)

(787, 30) (197, 30)
(787,) (197,)


Now we want to assign the Logistic regression algorithm as our training model.

In [52]:
model = LogisticRegression()

In [53]:
#Train the model
model.fit(X_train, Y_train)

LogisticRegression()

In [55]:
#the prediction for our X_train data
Xtrain_prediction = model.predict(X_train)

#compare our X_train predictions to our original Y_train labels
accuracy = accuracy_score(Xtrain_prediction, Y_train)

In [56]:
print("Accuracy is: ", accuracy)

Accuracy is:  0.9453621346886912


Our accuracy score is 0.94 which means there's a 94% chance the prediction is accurate. 

In [58]:
#make predictions from our test data
Xtest_prediction = model.predict(X_test)

#check accuracy of our precition to our test data
testAccuracy = accuracy_score(Xtest_prediction, Y_test)

In [59]:
print("Accuracy is: ", testAccuracy)

Accuracy is:  0.9593908629441624


Our test accuracy score is 0.95 which means there's a 95% chance the prediciton is accurate.